core/include/traccc/fitting/kalman_filter/two_filters_smoother.hpp

```diff
@@ -72,63 +74,70 @@ struct two_filters_smoother {
         const matrix_type<D, 1>& meas_local =
             trk_state.template measurement_local<D>();
 
+        // Spatial resolution (Measurement covariance)
+        const matrix_type<D, D> V =
+            trk_state.template measurement_covariance<D>();
+
         // Predicted vector of bound track parameters
-        const matrix_type<e_bound_size, 1> predicted_vec =
+        const matrix_type<e_bound_size, 1>& predicted_vec =
             bound_params.vector();
 
         // Predicted covaraince of bound track parameters
-        const matrix_type<e_bound_size, e_bound_size> predicted_cov =
+        const matrix_type<e_bound_size, e_bound_size>& predicted_cov =
             bound_params.covariance();
 
-        const matrix_type<e_bound_size, e_bound_size> predicted_cov_inv =
-            matrix::inverse(predicted_cov);
-        const matrix_type<e_bound_size, e_bound_size> filtered_cov_inv =
-            matrix::inverse(trk_state.filtered().covariance());
-
-        // Eq (3.38) of "Pattern Recognition, Tracking and Vertex
-        // Reconstruction in Particle Detectors"
-        const matrix_type<e_bound_size, e_bound_size> smoothed_cov_inv =
-            predicted_cov_inv + filtered_cov_inv;
-
-        const matrix_type<e_bound_size, e_bound_size> smoothed_cov =
-            matrix::inverse(smoothed_cov_inv);
-
-        // Eq (3.38) of "Pattern Recognition, Tracking and Vertex
-        // Reconstruction in Particle Detectors"
-        const matrix_type<e_bound_size, 1u> smoothed_vec =
-            smoothed_cov * (filtered_cov_inv * trk_state.filtered().vector() +
-                            predicted_cov_inv * predicted_vec);
-
-        trk_state.smoothed().set_vector(smoothed_vec);
-        trk_state.smoothed().set_covariance(smoothed_cov);
-
-        const matrix_type<D, 1> residual_smt = meas_local - H * smoothed_vec;
-
-        // Spatial resolution (Measurement covariance)
-        const matrix_type<D, D> V =
-            trk_state.template measurement_covariance<D>();
-
-        // Eq (3.39) of "Pattern Recognition, Tracking and Vertex
-        // Reconstruction in Particle Detectors"
-        const matrix_type<D, D> R_smt =
-            V - H * smoothed_cov * matrix::transpose(H);
-
-        // Eq (3.40) of "Pattern Recognition, Tracking and Vertex
-        // Reconstruction in Particle Detectors"
-        const matrix_type<1, 1> chi2_smt = matrix::transpose(residual_smt) *
-                                           matrix::inverse(R_smt) *
-                                           residual_smt;
+        // Do not smoothe if the forward pass produced an error
+        if (!trk_state.filtered().is_invalid()) {
```

Comment:
If the track state is invalid, should we not do something specific? I.e. should there not be some return type specific to that case?

Reply:
Done